memtable/write_buffer_manager.cc

```diff
@@ -0,0 +1,115 @@
+  char cache_key_prefix_[kCacheKeyPrefix];
+  uint64_t num_cache_entries_ = 0;
+  std::vector<Cache::Handle*> dummy_handles_;
+
+  Slice GetCacheKey(char* cache_key, uint64_t id) {
+    memcpy(cache_key, cache_key_prefix_, kCacheKeyPrefix);
+    char* end = EncodeVarint64(cache_key + kCacheKeyPrefix, id);
+    return Slice(cache_key, static_cast<size_t>(end - cache_key));
+  }
+};
+#else
+struct WriteBufferManager::CacheRep {};
+#endif  // ROCKSDB_LITE
+
+WriteBufferManager::WriteBufferManager(size_t _buffer_size,
+                                       std::shared_ptr<Cache> cache)
+    : buffer_size_(_buffer_size),
+      memory_used_(0),
+      mutable_memtable_used_(0),
+      cache_rep_(nullptr) {
+#ifndef ROCKSDB_LITE
+  if (cache) {
+    // Construct the cache key using the pointer to this.
+    cache_rep_.reset(new CacheRep());
+    cache_rep_->cache_ = cache;
+    memset(cache_rep_->cache_key_prefix_, 0, kCacheKeyPrefix);
+    size_t pointer_size = sizeof(const void*);
+    assert(pointer_size <= kCacheKeyPrefix);
+    memcpy(cache_rep_->cache_key_prefix_, static_cast<const void*>(this),
+           pointer_size);
+  }
+#endif  // ROCKSDB_LITE
+}
+
+WriteBufferManager::~WriteBufferManager() {
+#ifndef ROCKSDB_LITE
+  if (cache_rep_) {
+    for (auto* handle : cache_rep_->dummy_handles_) {
+      cache_rep_->cache_->Release(handle, true);
+    }
+  }
+#endif  // ROCKSDB_LITE
+}
+
+// Should only be called from write thread
+void WriteBufferManager::ReserveMemWithCache(size_t mem) {
+#ifndef ROCKSDB_LITE
+  assert(cache_rep_ != nullptr);
+  // Use a mutex to protect various data structures. Can be optimzied to a
+  // lock-free solution if it ends up with a performance bottleneck.
+  std::lock_guard<std::mutex> lock(cache_rep_->cache_mutex_);
+
+  size_t new_mem_used = memory_used_.load(std::memory_order_relaxed) + mem;
+  memory_used_.store(new_mem_used, std::memory_order_relaxed);
+  while (new_mem_used > cache_rep_->cache_allocated_size_) {
+    // Expand size by at least 1MB.
+    // Add a dummy record to the cache
+    char cache_key[kCacheKeyPrefix + kMaxVarint64Length];
+    Cache::Handle* handle;
+    cache_rep_->cache_->Insert(
+        cache_rep_->GetCacheKey(cache_key, cache_rep_->num_cache_entries_++),
+        nullptr, kSizeDummyEntry, nullptr, &handle);
+    cache_rep_->dummy_handles_.push_back(handle);
+    cache_rep_->cache_allocated_size_ += kSizeDummyEntry;
+  }
+#endif  // ROCKSDB_LITE
+}
+
+void WriteBufferManager::FreeMemWithCache(size_t mem) {
+#ifndef ROCKSDB_LITE
+  assert(cache_rep_ != nullptr);
+  // Use a mutex to protect various data structures. Can be optimzied to a
+  // lock-free solution if it ends up with a performance bottleneck.
+  std::lock_guard<std::mutex> lock(cache_rep_->cache_mutex_);
+  size_t new_mem_used = memory_used_.load(std::memory_order_relaxed) - mem;
+  memory_used_.store(new_mem_used, std::memory_order_relaxed);
+  if (new_mem_used < cache_rep_->cache_allocated_size_ * 3 / 4 &&
```

Comment:
It's an arbitrary number. I just want to make sure that
1. we don't pay the cost of the block cache immediately a memtable is freed, as it is expensive.
2. eventually, if we walk away from a temporary memtable size increase, we'll release those memory.
I come up with the 3/4 threshold to shrink back.